File: pipeline/graph.go
package main

// The pipeline graph. One rule governs the whole thing:
//
//	a node needs recomputing only if its own inputs changed, or something
//	upstream of it did.
//
// That is enforced by making a node's key include the keys of its
// dependencies (a merkle chain), so "upstream changed" propagates by
// construction instead of by anyone remembering to declare it. There is no
// second freshness mechanism and no special case for the shadcn pin: the pin
// file is simply an input of the node that reads it, and its key flows
// downstream from there.

import (
	"fmt"
	"os"
)

type Node struct {
	ID       NodeID
	Kind     string
	Tier     string
	Needs    []NodeID
	Run      [][]string
	Inputs   []string // nil = judges state outside the tree, never fresh
	Produces []string

	// Why is one line: what breaks if this node is deleted. Not decoration —
	// TestMeta requires every gate to carry one, and the runner prints it when
	// the node goes red.
	Why string
	// Mutations are ids under gates/mutations/ that MUST make this gate fail.
	// A gate with no mutation is unproven and TestMeta rejects it.
	Mutations []string
}

// NeverFresh reports whether the node declares no input set. Such a node
// cannot be skipped and neither can anything downstream of it.
func (n Node) NeverFresh() bool { return n.Inputs == nil }

type Graph struct {
	nodes map[NodeID]Node
	order []NodeID // declaration order, for stable output
}

// LoadGraph builds the graph rooted at the working directory. The runner runs
// with the repo root as its cwd; anything else (a test, a tool) should say
// which root it means with LoadGraphAt.
func LoadGraph() (*Graph, error) {
	root, err := os.Getwd()
	if err != nil {
		return nil, err
	}
	return LoadGraphAt(root)
}

// LoadGraphAt builds the graph from Nodes, expanding any node the fan-out
// table splits into independent per-item nodes (which needs the tree, hence
// the root).
func LoadGraphAt(root string) (*Graph, error) {
	list, err := expandFanout(root, Nodes)
	if err != nil {
		return nil, err
	}
	return newGraph(list)
}

// AuthoredGraph is the graph exactly as declared in nodes.go, with no fan-out
// applied. This is the graph a human reviews and the one mutations name: a
// mutation targets the gate `contracts`, which the runner splits into
// `contracts:dialog` and 28 siblings. Meta works at this level so a mutation
// keeps naming the reviewable node rather than one arbitrary shard of it.
func AuthoredGraph() (*Graph, error) { return newGraph(Nodes) }

func newGraph(list []Node) (*Graph, error) {
	g := &Graph{nodes: make(map[NodeID]Node, len(list))}
	for _, n := range list {
		if _, dup := g.nodes[n.ID]; dup {
			return nil, fmt.Errorf("duplicate node id: %s", n.ID)
		}
		g.nodes[n.ID] = n
		g.order = append(g.order, n.ID)
	}
	for _, n := range list {
		for _, d := range n.Needs {
			if _, ok := g.nodes[d]; !ok {
				return nil, fmt.Errorf("node %s needs unknown node %s", n.ID, d)
			}
		}
	}
	return g, nil
}

func (g *Graph) Node(id NodeID) (Node, bool) { n, ok := g.nodes[id]; return n, ok }
func (g *Graph) IDs() []NodeID               { return g.order }

// Plan returns the transitive closure of targets, topologically sorted.
// Cycles are an authoring error and are reported with the path that closed them.
func (g *Graph) Plan(targets []NodeID) ([]Node, error) {
	seen := map[NodeID]bool{}
	visiting := map[NodeID]bool{}
	var out []Node
	var visit func(id NodeID, path []NodeID) error
	visit = func(id NodeID, path []NodeID) error {
		if seen[id] {
			return nil
		}
		if visiting[id] {
			return fmt.Errorf("cycle in graph: %v", append(path, id))
		}
		n, ok := g.nodes[id]
		if !ok {
			return fmt.Errorf("unknown node id: %s", id)
		}
		visiting[id] = true
		for _, d := range n.Needs {
			if err := visit(d, append(path, id)); err != nil {
				return err
			}
		}
		delete(visiting, id)
		seen[id] = true
		out = append(out, n)
		return nil
	}
	for _, t := range targets {
		if err := visit(t, nil); err != nil {
			return nil, err
		}
	}
	return out, nil
}

// Two rungs, not three. "medium" meant "compiles, no browser" and existed for
// emit-smoke, the only gate that ever declared it; with that gate gone nothing
// selected the rung and `run medium` returned the fast set. The nodes that
// declared it are now "full" — which changes no plan, because no fast gate
// reaches them and the full tier already did.
var tiers = []string{"fast", "full"}

func tierRank(t string) int {
	for i, x := range tiers {
		if x == t {
			return i
		}
	}
	return len(tiers) // unknown tier sorts last rather than silently passing as fast
}

// EffectiveTier is the most expensive tier in a node's dependency closure,
// not the tier it declares: a "fast" gate that can only run after a browser
// build is not fast.
func (g *Graph) EffectiveTier(id NodeID) string {
	memo := map[NodeID]int{}
	var walk func(NodeID) int
	walk = func(id NodeID) int {
		if r, ok := memo[id]; ok {
			return r
		}
		n := g.nodes[id]
		worst := tierRank(n.Tier)
		memo[id] = worst // cycle guard; Plan reports real cycles
		for _, d := range n.Needs {
			if r := walk(d); r > worst {
				worst = r
			}
		}
		memo[id] = worst
		return worst
	}
	return tiers[walk(id)]
}

// PlanTier is every gate whose ENTIRE closure fits within tier, plus the
// builds those gates need.
func (g *Graph) PlanTier(tier string) ([]Node, error) {
	max := tierRank(tier)
	var targets []NodeID
	for _, id := range g.order {
		if g.nodes[id].Kind == "gate" && tierRank(g.EffectiveTier(id)) <= max {
			targets = append(targets, id)
		}
	}
	return g.Plan(targets)
}

// PlanBuilds is every build node and what it needs — every artifact, no gates.
// gates/meta.mjs wants exactly this as its prelude: mutation testing needs the
// tree built, but running the gates first would just be running them twice.
func (g *Graph) PlanBuilds() ([]Node, error) {
	var targets []NodeID
	for _, id := range g.order {
		if g.nodes[id].Kind == "build" {
			targets = append(targets, id)
		}
	}
	return g.Plan(targets)
}

// runInputs prints the files a node's patterns actually resolve to. The
// globs are prose until you see what they expand to — this is what makes a
// declaration reviewable, and it is what the sandbox builds its tree from.
func runInputs(args []string) int {
	root, err := os.Getwd()
	if err != nil {
		fmt.Fprintln(os.Stderr, "pipeline:", err)
		return 1
	}
	var id string
	produces := false
	for _, a := range args {
		if a == "--produces" {
			produces = true
			continue
		}
		id = a
	}
	if id == "" {
		fmt.Fprintln(os.Stderr, "usage: pipeline inputs <node> [--produces]")
		return 2
	}
	g, err := LoadGraphAt(root)
	if err != nil {
		fmt.Fprintln(os.Stderr, "pipeline:", err)
		return 1
	}
	n, ok := g.Node(NodeID(id))
	if !ok {
		fmt.Fprintf(os.Stderr, "unknown node: %s\n", id)
		return 2
	}
	patterns := n.Inputs
	if produces {
		patterns = n.Produces
	}
	if patterns == nil {
		return 0 // never-fresh: no declared set at all
	}
	files, err := Files(root, patterns)
	if err != nil {
		fmt.Fprintln(os.Stderr, "pipeline:", err)
		return 1
	}
	for _, f := range files {
		fmt.Println(f)
	}
	return 0
}

File: pipeline/src/graph.rs
//! Port of pipeline/graph.go — the pipeline graph and the one rule:
//!
//! ```text
//! a node needs recomputing only if its own inputs changed, or something
//! upstream of it did.
//! ```
//!
//! `plan` is a declaration-order DFS post-order (targets in argv order, each
//! node's `needs` in declared order) — the traversal order IS the output
//! interface and is golden-diffed (tests/gen_golden.sh records this engine's
//! own bytes).

use crate::nodes::{self, Node};
use std::collections::{HashMap, HashSet};
use std::path::Path;

const TIERS: [&str; 2] = ["fast", "full"];

fn tier_rank(t: &str) -> usize {
    for (i, x) in TIERS.iter().enumerate() {
        if *x == t {
            return i;
        }
    }
    TIERS.len() // unknown tier sorts last rather than silently passing as fast
}

pub struct Graph {
    nodes: HashMap<String, Node>,
    order: Vec<String>, // declaration order, for stable output
}

impl Graph {
    pub(crate) fn new(list: Vec<Node>) -> Result<Graph, String> {
        let mut g = Graph {
            nodes: HashMap::with_capacity(list.len()),
            order: Vec::with_capacity(list.len()),
        };
        for n in list {
            if g.nodes.contains_key(&n.id) {
                return Err(format!("duplicate node id: {}", n.id));
            }
            g.order.push(n.id.clone());
            g.nodes.insert(n.id.clone(), n);
        }
        for n in g.nodes.values() {
            for d in &n.needs {
                if !g.nodes.contains_key(d) {
                    return Err(format!("node {} needs unknown node {}", n.id, d));
                }
            }
        }
        Ok(g)
    }

    pub fn node(&self, id: &str) -> Option<&Node> {
        self.nodes.get(id)
    }

    pub fn ids(&self) -> &[String] {
        &self.order
    }

    /// The transitive closure of targets, in declaration-order DFS post-order.
    /// Cycles are an authoring error, reported with the path that closed them.
    pub fn plan(&self, targets: &[String]) -> Result<Vec<Node>, String> {
        let mut seen = HashSet::new();
        let mut visiting = HashSet::new();
        let mut out = Vec::new();
        for t in targets {
            visit(self, t, &[], &mut seen, &mut visiting, &mut out)?;
        }
        Ok(out)
    }

    /// The most expensive tier in a node's dependency closure, not the tier it
    /// declares: a "fast" gate that can only run after a browser build is not fast.
    pub fn effective_tier(&self, id: &str) -> String {
        fn walk(g: &Graph, id: &str, memo: &mut HashMap<String, usize>) -> usize {
            if let Some(r) = memo.get(id) {
                return *r;
            }
            let n = g.node(id).expect("graph is closed over needs");
            let mut worst = tier_rank(&n.tier);
            memo.insert(id.to_string(), worst); // cycle guard; Plan reports real cycles
            for d in &n.needs {
                let r = walk(g, d, memo);
                if r > worst {
                    worst = r;
                }
            }
            memo.insert(id.to_string(), worst);
            worst
        }
        let mut memo = HashMap::new();
        TIERS[walk(self, id, &mut memo)].to_string()
    }

    /// Every gate whose ENTIRE closure fits within tier, plus the builds those
    /// gates need.
    pub fn plan_tier(&self, tier: &str) -> Result<Vec<Node>, String> {
        let max = tier_rank(tier);
        let targets: Vec<String> = self
            .order
            .iter()
            .filter(|id| {
                let n = self.nodes.get(*id).expect("order holds only graph nodes");
                n.kind == "gate" && tier_rank(&self.effective_tier(id)) <= max
            })
            .cloned()
            .collect();
        self.plan(&targets)
    }

    /// Every build node and what it needs — every artifact, no gates.
    pub fn plan_builds(&self) -> Result<Vec<Node>, String> {
        let targets: Vec<String> = self
            .order
            .iter()
            .filter(|id| self.nodes.get(*id).expect("order nodes").kind == "build")
            .cloned()
            .collect();
        self.plan(&targets)
    }
}

fn visit(
    g: &Graph,
    id: &str,
    path: &[String],
    seen: &mut HashSet<String>,
    visiting: &mut HashSet<String>,
    out: &mut Vec<Node>,
) -> Result<(), String> {
    if seen.contains(id) {
        return Ok(());
    }
    if visiting.contains(id) {
        let mut p = path.to_vec();
        p.push(id.to_string());
        return Err(format!("cycle in graph: [{}]", p.join(" ")));
    }
    let n = g.node(id).ok_or_else(|| format!("unknown node id: {}", id))?;
    visiting.insert(id.to_string());
    let mut child_path = path.to_vec();
    child_path.push(id.to_string());
    for d in &n.needs {
        visit(g, d, &child_path, seen, visiting, out)?;
    }
    visiting.remove(id);
    seen.insert(id.to_string());
    out.push(n.clone());
    Ok(())
}

/// LoadGraphAt: build the graph from nodes::all(), expanding any node the
/// fan-out table splits into independent per-item nodes and resolving the
/// `produces` entries that are derived from the tree rather than written out.
pub fn load_graph_at(root: &Path) -> Result<Graph, String> {
    let list = crate::fanout::expand_fanout(root, nodes::all())?;
    let list = crate::produces::apply_derived_produces(root, list)?;
    Graph::new(list)
}

/// AuthoredGraph (Go meta.go): the authored table with no fanout expansion —
/// the view the mutation harness reasons about (`contracts` as one gate, not
/// 29 shards).
pub fn authored() -> Result<Graph, String> {
    Graph::new(nodes::all())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn authored() -> Graph {
        Graph::new(nodes::all()).expect("authored graph is consistent")
    }

    #[test]
    fn authored_graph_shape() {
        let g = authored();
        assert_eq!(g.ids().len(), 41);
        assert!(g.node("pin").is_some());
        assert!(g.node("contracts").is_some());
        assert!(g.node("nope").is_none());
    }

    #[test]
    fn plan_is_declaration_order_dfs_post_order() {
        let g = authored();
        // "pin" is declared first and needs nothing: it is the whole plan.
        let plan = g.plan(&["pin".to_string()]).unwrap();
        assert_eq!(plan.len(), 1);
        assert_eq!(plan[0].id, "pin");
        // example-oracle needs emit then build-js (declared order): build-js's
        // own needs (none) come first, then emit's closure (convert's needs
        // pin then build-js), so pin precedes build-js even though targets
        // start at example-oracle.
        let plan = g.plan(&["example-oracle".to_string()]).unwrap();
        let ids: Vec<&str> = plan.iter().map(|n| n.id.as_str()).collect();
        let pos = |s: &str| ids.iter().position(|x| *x == s).unwrap();
        assert!(pos("pin") < pos("build-js"));
        assert!(pos("build-js") < pos("convert"));
        assert!(pos("convert") < pos("emit"));
        assert!(pos("emit") < pos("example-oracle"));
        assert_eq!(*ids.last().unwrap(), "example-oracle");
    }

    #[test]
    fn cycle_reports_the_closing_path() {
        let mut n_a = nodes::all().into_iter().find(|n| n.id == "pin").unwrap();
        n_a.needs = vec!["unit".to_string()];
        let mut list = nodes::all();
        list.retain(|n| n.id != "pin");
        list.push(n_a);
        // unit needs build-js; rewire pin->unit and unit->pin to close a cycle
        let unit = list.iter_mut().find(|n| n.id == "unit").unwrap();
        unit.needs.push("pin".to_string());
        let g = Graph::new(list).unwrap();
        let err = g.plan(&["pin".to_string()]).unwrap_err();
        assert_eq!(err, "cycle in graph: [pin unit pin]");
    }

    #[test]
    fn duplicate_and_unknown_deps_are_authoring_errors() {
        let mut list = nodes::all();
        let pin = list.first().unwrap().clone();
        list.push(pin);
        let err = Graph::new(list.clone()).err().unwrap();
        assert_eq!(err, "duplicate node id: pin");
        let mut list = nodes::all();
        list.first_mut().unwrap().needs.push("ghost".to_string());
        let err = Graph::new(list).err().unwrap();
        assert_eq!(err, "node pin needs unknown node ghost");
    }

    #[test]
    fn effective_tier_is_the_closure_maximum() {
        let g = authored();
        assert_eq!(g.effective_tier("pin"), "fast");
        // css-direction declares fast but sits downstream of demo (full).
        assert_eq!(g.effective_tier("css-direction"), "full");
        assert_eq!(g.effective_tier("typecheck"), "fast");
    }

    #[test]
    fn plan_tier_fast_is_every_fast_gate_plus_needed_builds() {
        let g = authored();
        let plan = g.plan_tier("fast").unwrap();
        let gates: Vec<&str> = plan
            .iter()
            .filter(|n| n.kind == "gate")
            .map(|n| n.id.as_str())
            .collect();
        // gates whose ENTIRE closure fits in fast: unit only needs build-js
        // (a fast build) so it stays; dist-complete/pack/coverage/overlay/
        // css-direction and both docs gates all reach a full-tier build.
        // docs-upstream-mirror is tier fast but kind build — not a gate.
        assert_eq!(
            gates,
            vec!["pin", "unit", "typecheck", "ledger", "script-refs"]
        );
        assert!(plan.iter().all(|n| n.tier == "fast" || n.kind == "build"));
    }

    #[test]
    fn plan_builds_has_no_gates() {
        let g = authored();
        let plan = g.plan_builds().unwrap();
        // the closure is "every build node and what it needs": pin (a gate)
        // legitimately appears because convert/rtl-dict/docs-upstream-mirror
        // need it; every other entry is a build.
        assert!(plan.iter().all(|n| n.kind == "build" || n.id == "pin"));
        assert!(plan.iter().any(|n| n.id == "emit"));
        assert!(plan.iter().any(|n| n.id == "pin"));
        assert!(!plan.iter().any(|n| n.id == "coverage"));
    }
}
